### PR TITLE
[W11H03] DataHandler tests

### PR DESCRIPTION
**Pull request by felixhauptmann**
This tests if the relevant methods send the correct request and produce the expected output/side-effects when met with different responses.

P.S. feel free to fork and add test cases.

### CODE REVIEW

**Review by Anatoly03 — APPROVED** [review]
lgtm!

File: w11h03/test/pgdp/networking/DataHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+    @AfterEach
+    public void cleanup() {
+        server.stop(0);
+    }
+
+    void registerRequestTest(Request actual, String username, String id) {
+        assertEquals("POST", actual.requestMethod());
+        assertEquals(List.of("application/json"), actual.requestHeaders().get("Content-Type"));
+        assertEquals(URI.create("http://carol.sse.cit.tum.de/api/users/register"), actual.requestURI());
```
Comment (faid0) [inline w11h03/test/pgdp/networking/DataHandlerTest.java:62]:
```suggestion
        assertEquals(URI.create("http://carol.sse.cit.tum.de/api/user/register"), actual.requestURI());
```
see http://131.159.47.114/docs

File: w11h03/test/pgdp/networking/DataHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        assertEquals("{\"username\": \"" + username + "\",\"tum_kennung\": \"" + id + "\"}", actual.getBodyUTF8());
+    }
+
+    @Test
+    public void registerSuccessTest() {
+        Pair<Boolean, Request> result = testEndpoint("/api/users/register", 200, "", () -> dataHandler.register(USERNAME, ID)); // TODO: https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/Definition.20erfolgreiche.20Anfragen
```
Comment (faid0) [inline w11h03/test/pgdp/networking/DataHandlerTest.java:68]:
```suggestion
        Pair<Boolean, Request> result = testEndpoint("/api/user/register", 200, "", () -> dataHandler.register(USERNAME, ID)); // TODO: https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/Definition.20erfolgreiche.20Anfragen
```
see above

File: w11h03/test/pgdp/networking/DataHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        registerRequestTest(result.getRight(), USERNAME, ID);
+        assertTrue(result.getLeft());
+    }
+
+    @Test
+    public void registerFailTest() {
+        Pair<Boolean, Request> result = testEndpoint("/api/users/register", 422, "", () -> dataHandler.register(USERNAME, ID)); // TODO: https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/Definition.20erfolgreiche.20Anfragen
```
Comment (faid0) [inline w11h03/test/pgdp/networking/DataHandlerTest.java:75]:
```suggestion
        Pair<Boolean, Request> result = testEndpoint("/api/user/register", 422, "", () -> dataHandler.register(USERNAME, ID)); // TODO: https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/Definition.20erfolgreiche.20Anfragen
```
see above

**Review by faid0 — CHANGES_REQUESTED** [review]
wrong api endpoint

**Review by Anatoly03 — APPROVED** [review]
lgtm!

File: w11h03/test/pgdp/networking/DataHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            exchange.getResponseBody().write(body);
+            exchange.close();
+        });
+
+        T result = fn.get();
+
+        server.removeContext(endpoint);
+
+        return Pair.of(result, Objects.requireNonNull(req.get(), "provided function did not send request to specified endpoint"));
+    }
```
Comment (Anatoly03) [inline w11h03/test/pgdp/networking/DataHandlerTest.java:116]: could you make this method public so other files can use it too?

Comment (felixhauptmann) [reply w11h03/test/pgdp/networking/DataHandlerTest.java]: Latest commit should more or less resolve this

File: w11h03/test/pgdp/networking/DataHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import java.net.URI;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static pgdp.networking.Util.*;
+
+public class DataHandlerTest {
+
+    private static final String USERNAME = "TestUser123";
+    private static final String ID = "geXYZtup";
```
Comment (Anatoly03) [inline w11h03/test/pgdp/networking/DataHandlerTest.java:16]:
```suggestion
    private static final String ID = "ge69sex";
```
it’s important to follow digits and and the ending of format consonant vowel consonant

**Review by Anatoly03 — APPROVED** [review]
lgtm!

File: w11h03/test/pgdp/networking/DataHandlerTest.java (hunk 5 above)
Comment (felixhauptmann) [reply w11h03/test/pgdp/networking/DataHandlerTest.java:16]: If you say so xD

**Review by nilsreichardt — APPROVED** [review]
LGTM

**Pull request merged**